### ISSUE TITLE
Improve docs for testing source generators

### ISSUE THREAD

**Issue by giggio** [dotnet/roslyn#58945]
The current docs for tests for source generators is missing some important information about testing. For example, how to add references to the compilation. The current example is a good quick start but is not enough.

Docs link:
https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#unit-testing-of-generators

**Comment by AlbertoMonteiro** [comment]
When I was working with a Source Generator that requires AspNetCore 6.0 I stuck for hours to find how to make the `CSharpSourceGeneratorTest` be able to know AspNetCore assemblies/types.

```csharp
new ReferenceAssemblies("net6.0", new("Microsoft.NETCore.App.Ref", "6.0.0"), @"ref\net6.0")
     .WithPackages(ImmutableArray.Create(new PackageIdentity("Microsoft.AspNetCore.App.Ref", "6.0.0")));
```
After some hours and many tries, I tried "Go To Definition" of the WebApplication class and I ended up getting the package name from that code below
![image](https://user-images.githubusercontent.com/836496/150141976-5fbd5ec1-e88c-47dc-88ea-7f101beadd8b.png)

Being stuck to write that small piece of code(`new PackageIdentity("Microsoft.AspNetCore.App.Ref", "6.0.0")`) is really annoying

**Comment by Youssef1313** [comment]
> new ReferenceAssemblies("net6.0", new("Microsoft.NETCore.App.Ref", "6.0.0"), @"ref\net6.0")

You don't need to do that I believe. You can use the already existing property `Microsoft.CodeAnalysis.Testing.ReferenceAssemblies.Net.Net60`.

**Comment by Youssef1313** [comment]
I think this is not source generators specific. It's a question about the testing library in general. It might worth documenting in dotnet/docs repository for wider visibility. @BillWagner @IEvangelist What do you think?

**Comment by AlbertoMonteiro** [comment]
> > new ReferenceAssemblies("net6.0", new("Microsoft.NETCore.App.Ref", "6.0.0"), @"ref\net6.0")
> 
> You don't need to do that I believe. You can use the already existing property `Microsoft.CodeAnalysis.Testing.ReferenceAssemblies.Net.Net60`.

There is not Net60
![image](https://user-images.githubusercontent.com/836496/150154894-416e7959-fe19-4cb1-91c7-2e857728380d.png)

**Comment by Youssef1313** [comment]
@AlbertoMonteiro You're probably consuming an old version of the testing library.

**Comment by AlbertoMonteiro** [comment]
@Youssef1313 actually I am using the latest version 

https://www.nuget.org/packages/Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing.XUnit/

![image](https://user-images.githubusercontent.com/836496/150155730-34231b70-9480-4084-8e9b-86e44bd5b9d4.png)

![image](https://user-images.githubusercontent.com/836496/150155826-8f22fe63-611c-4b5b-a1c6-b82ca7f96f57.png)

**Comment by Youssef1313** [comment]
Hmm, it looks like the version with Net60 available is only released in the Azure DevOps feed, but not on nuget.org feed.

@jmarolf @sharwell Can a next version be released on nuget.org?

**Comment by sharwell** [comment]
@AlbertoMonteiro the version on nuget.org will only be "current" for a few weeks at the very most. I recommend using the one from this feed:

https://github.com/dotnet/roslyn/blob/259a9e27ab8c6a1d8d3a3445b079fd5e3a13ec0a/NuGet.config#L7

**Comment by giggio** [comment]
Why not just publish the new version to Nuget.org? Is it unstable or not finished?

**Comment by giggio** [comment]
The problem with docs remain, BTW, even when the new lib version is published to Nuget.org.

**Comment by AlbertoMonteiro** [comment]
I agree with @giggio, I spent more time trying to figure out how to reference aspnet in the test context than anything else, so basically almost 2~3 hours to discover that code `new PackageIdentity("Microsoft.AspNetCore.App.Ref", "6.0.0")`.
So yeah, a new nuget package version won't fix that issue of lacking information about how to do those things.

**Comment by sharwell** [comment]
> Why not just publish the new version to Nuget.org? Is it unstable or not finished?

It's not an automated process, and it's always going to lag behind the daily builds to some degree. Since this dependency is only consumed by test projects, there's no significant consumer downside for referencing the dotnet-tools feed (aside from the one-time setup) or referencing a beta build.

**Comment by AraHaan** [comment]
CSharpSourceGeneratorTest also lacks being able to test incremental generators which I think is a must have.

**Comment by juwens** [comment]
> CSharpSourceGeneratorTest also lacks being able to test incremental generators which I think is a must have.
> 
> Note: It's easy to create a special `CSharpIncrementalGeneratorTest` class to handle incremental generators. 😄

can you please provide this "simple" implementation on a gist (or so). Other people not writing source generators every day, might not find it not as easy as you.

**Comment by grosch-intl** [comment]
> CSharpSourceGeneratorTest also lacks being able to test incremental generators which I think is a must have.
> 
> Note: It's easy to create a special `CSharpIncrementalGeneratorTest` class to handle incremental generators. 😄

@AraHaan Can you please provide an example?  The cookbook says to use that class, but nothing says how to create it or what it should contain.

**Comment by AraHaan** [comment]
Sure, I will see if I can dig that code back up again from one of my repositories. @grosch-intl.

**Comment by AraHaan** [comment]
Upon looking back, it looks like I never commited the files and hopefully they are on my macbook to where I can push them to github.

**Comment by AraHaan** [comment]
I found the code, it's all in this folder: https://github.com/Elskom/IDisposableGenerator/tree/main/tests

**Comment by sharwell** [comment]
> CSharpSourceGeneratorTest also lacks being able to test incremental generators

This is outdated information (the `ISourceGenerator` constraint was removed in https://github.com/dotnet/roslyn-sdk/pull/1084), so I'm resolving these comments as outdated to collapse them in this view.
